### PR TITLE
[SYCL] Handle captures by ref in kernel lambda

### PR DESCRIPTION
**Pull request by ilyastepykin**
Captures by reference wasn't properly handled
by the compiler which led to a crash.

References are not standard layout so if we
see one, emit a corresponding error as for other
non-standard layout structures/classes.

Signed-off-by: Ilya Stepykin <ilya.stepykin@intel.com>

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -1054,17 +1055,19 @@ static void buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
...
-        KernelObj->getASTContext().getDiagnostics().Report(
-            V->getLocation(), diag::err_sycl_non_std_layout_type);
-      }
+      const DeclaratorDecl *V =
+          Cpt ? cast<DeclaratorDecl>(Cpt->getCapturedVar())
+              : cast<DeclaratorDecl>(Fld);
+      KernelObj->getASTContext().getDiagnostics().Report(
+          V->getLocation(), diag::err_sycl_non_std_layout_type);
+      return false;
+    } else if (ArgTy->isStructureOrClassType()) {
```
Comment (bader) [inline clang/lib/Sema/SemaSYCL.cpp]: Please, don't use else after return.
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

File: clang/test/SemaSYCL/non-std-layout-param.cpp [hunk 2]
```diff
@@ -26,3 +26,16 @@ void test() {
...
 }
 
+void test_capture_single_ref() {
+  // expected-error@+1 {{kernel parameter has non-standard layout class/struct type}}
+  int p = 0;
+  kernel_single_task<class kernel_capture_single_ref>([&p] { (void) p; });
+}
+
+void test_capture_refs() {
+  // Only one error as we stop processing after first non-standard layout parameter
```
Comment (bader) [inline clang/test/SemaSYCL/non-std-layout-param.cpp]: Why do we stop?

Comment (ilyastepykin) [reply clang/test/SemaSYCL/non-std-layout-param.cpp]: Looks it's better to go though all the parameters to emit a better diagnostic, update the code to do that.

Comment (bader) [reply clang/test/SemaSYCL/non-std-layout-param.cpp]: > Looks it's better to go though all the parameters to emit a better diagnostic.

Right.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -1077,14 +1085,21 @@ static void buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
...
                                            Quals);
       QualType ModTy = Context.getPointerType(PointeeTy);
-      
+
       CreateAndAddPrmDsc(Fld, ModTy);
     } else if (ArgTy->isScalarType()) {
       CreateAndAddPrmDsc(Fld, ArgTy);
     } else {
       llvm_unreachable("Unsupported kernel parameter type");
+      return false;
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp]: What happened here?  This seems like an unrelated change?

Comment (ilyastepykin) [reply clang/lib/Sema/SemaSYCL.cpp]: It's unnecessary here, removed. Thanks.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -1054,17 +1056,23 @@ static void buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
...
-            Cpt ? cast<DeclaratorDecl>(Cpt->getCapturedVar())
-                : cast<DeclaratorDecl>(Fld);
-        KernelObj->getASTContext().getDiagnostics().Report(
-            V->getLocation(), diag::err_sycl_non_std_layout_type);
-      }
+      const auto& DiagLocation = Cpt ?
+          Cpt->getLocation() :
+          cast<DeclaratorDecl>(Fld)->getLocation();
+
+      KernelObj->getASTContext().getDiagnostics().Report(
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp]: Why is this going through KernelObj?  We already have the Context as a parameter, right?

Comment (ilyastepykin) [reply clang/lib/Sema/SemaSYCL.cpp]: Now sure why it was done so. Replaced with Context parameter as they're the same. Thanks.

**Comment by erichkeane** [comment]
I'm OK with this.  It won't let me review it though :?

File: clang/test/SemaSYCL/non-std-layout-param.cpp [hunk 5]
```diff
@@ -20,9 +20,61 @@ __attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
...
+    (void) q;
+    (void) p;
+    (void) s;
+  });
 }
 
+void test_capture_implicit_refs() {
+  int p = 0;
+  double q = 0;
+  kernel_single_task<class kernel_capture_refs>([&] {
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/non-std-layout-param.cpp]: Actually this case is prohibited by the 1.2.1. SYCL spec:
> The kernel function for the lambda function is the lambda function itself. The kernel lambda must use copy for
all of its captures (i.e. [=]).

Shouldn't we mention it in a diagnostic?

I also wonder which problem motivated us to made this change.

Comment (keryell) [reply clang/test/SemaSYCL/non-std-layout-param.cpp]: Probably this is required to produce some values to a USM pointer or other extensions.
